elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchHSQueryImpl.java

```diff
@@ -503,12 +495,108 @@ private void addScriptFields(JsonBuilder.Object query) {
 			}
 		}
 
-		SearchResult runSearch() {
+		SearchResult search() {
+			return search( false );
+		}
+
+		SearchResult searchWithScrollEnabled() {
+			return search( true );
+		}
+
+		private SearchResult search(boolean enableScrolling) {
+			Search.Builder builder = new Search.Builder( completeQueryAsString );
+			builder.addIndex( indexNames );
+
+			if ( enableScrolling ) {
+				/*
+				 * Note: "firstResult" is currently being ignored by Elasticsearch when scrolling.
+				 * See https://github.com/elastic/elasticsearch/issues/9373
+				 * To work this around, we don't use the "from" parameter here, and the document
+				 * extractor will skip the results by scrolling until it gets the right index.
+				 */
+
+				builder.setParameter( Parameters.SCROLL, getScrollTimeout() );
+
+				/*
+				 * The "size" parameter has a slightly different meaning when scrolling: it defines the window
+				 * size for the first search *and* for the next calls to the scroll API.
+				 * We still reduce the number of results in accordance to "maxResults", but that's done on our side
+				 * in the document extractor.
+				 */
+				builder.setParameter( Parameters.SIZE, getScrollFetchSize() );
+			}
+			else {
+				builder.setParameter( "from", firstResult );
+
+				// If the user has given a value, take it as is, let ES itself complain if it's too high; if no value is
+				// given, I take as much as possible, as by default only 10 rows would be returned
+				builder.setParameter( Parameters.SIZE, maxResults != null ? maxResults : MAX_RESULT_WINDOW_SIZE - firstResult );
+			}
+
+			// TODO: HSEARCH-2254 embedded fields (see https://github.com/searchbox-io/Jest/issues/304)
+			if ( sort != null ) {
+				validateSortFields( extendedIntegrator, getQueriedEntityTypes() );
+				for ( SortField sortField : sort.getSort() ) {
+					builder.addSort( getSort( sortField, idFieldName ) );
+				}
+			}
+			Search search = builder.build();
 			try ( ServiceReference<JestClient> client = getExtendedSearchIntegrator().getServiceManager().requestReference( JestClient.class ) ) {
 				return client.get().executeRequest( search );
 			}
 		}
 
+		private String getScrollTimeout() {
```

Comment:
Right, I totally agree. I did it this way because I couldn't see any central place where I could provide accessors for these properties. So sure, I'll see to that later :)